_overviews/scala3-book/domain-modeling-tools.md

```diff
@@ -0,0 +1,575 @@
+---
+title: Tools
+type: section
+description: This chapter provides an introduction to the available domain modeling tools in Scala 3, including classes, traits, enums, and more.
+num: 20
+previous-page: domain-modeling-intro
+next-page: domain-modeling-oop
+---
+
+
+
+Scala 3 provides the following tools to help us model the world around us:
+
+- Class
+- Companion object
+- Trait
+- Abstract class
+- Enum
+- Case class
+- Case object
+
+The following sections introduce each of these modeling tools.
+
+
+
+## Classes
+
+As with other languages, a Scala *class* is a template for the creation of object instances. Here are some examples of classes that have constructor parameters, but no fields or methods:
+
+```scala
+class Person(var name: String, var vocation: String)
+class Book(var title: String, var author: String, var year: Int)
+class Movie(var name: String, var director: String, var year: Int)
+```
+
+All of those parameters are defined as `var` fields, which means they are mutable: you can read them, and also modify them. If you want them to be immutable — read only — create them as `val` fields instead.
+
+Prior to Scala 3, you used the `new` keyword to create a new instance of a class:
+
+```scala
+val p = new Person("Robert Allen Zimmerman", "Harmonica Player")
+        ---
+```
+
+However, this isn’t required in Scala 3:
+
+```scala
+val p = new Person("Robert Allen Zimmerman", "Harmonica Player")
+```
+
+Once you have an instance of a class, you can access its fields, which in this example are all constructor parameters:
+
+```scala
+p.name       // "Robert Allen Zimmerman"
+p.vocation   // "Harmonica Player"
+```
+
+As mentioned, all of these parameters were created as `var` fields, so you can also mutate them:
+
+```scala
+p.name = "Bob Dylan"
+p.vocation = "Musician"
+```
+
+
+### Fields and methods
+
+In addition to constructor parameters, classes can also have fields and methods. They are defined in the body of the class, which also serves as the class constructor:
+
+```scala
+class Person(var firstName: String, var lastName: String):
+  
+  println("the constructor begins")
+  val fullName = firstName + " " + lastName
+
+  // a class method
+  def printFullName: Unit =
+    // access the `fullName` field, which is created above
+    println(fullName)
+  
+  printFullName
+  println("this ends the constructor")
+```
+
+The REPL shows how to create a new `Person` instance with this class:
```

Comment:
```suggestion
This REPL extract shows how to create a new `Person` instance with this class:
```

Reply:
maybe you mean "excerpt" instead of "extract"?